exist-core/src/main/java/org/exist/storage/NativeBroker.java

```diff
@@ -3171,9 +3171,9 @@ public Object start() {
                     return null;
                 }
             }.run();
-            // create a copy of the old doc to copy the nodes into it
+            // Create a copy of the old document to copy the nodes into it.
+            // tempDoc serves as a storage container for the child nodes.
             final DocumentImpl tempDoc = new DocumentImpl(null, doc.getDocId(), doc);
-            tempDoc.copyOf(this, doc, doc);
```

Comment:
I don't think just removing this is the correct or secure solution. Can you write a bit more detail in the description of the PR about exactly what the problem was please, that might help me to suggest a better solution.

Reply:
The commit message body has some more detail.

> With this patch applied, this can no longer happen as the call to DocumentImpl#copyOf is removed in defragXMLresource.
> The temporary document is just a container to realign nodes on one BTree page and is destroyed within the method.

I hope this clarifies the motivation and justification.
I would like to point out that we arrived at this solution together with @wolfgangmm

Reply:
> Likely needs a more nuanced fix

Please only block PRs where you have either proof that it creates a problem or at least share the thoughts what the issue might be.

Reply:
> Please only block PRs where you have either proof that it creates a problem

Obviously!

> or at least share the thoughts what the issue might be

I am concerned that the temporary document which looks persistent to me is not correctly initialised, also I am concerned that the security on it is now too relaxed.

Can you clearly describe the initial problem that you are trying to solve in the PR description please. For example, something like: There is an active user `U` with the security profile `s(U)`, there is document `D` with the permissions `p(D)`, when the action `A` is  executed by the subject `AS` with security profile `s(AS)`, it is rejected because `s(AS)` is incompatible with `p(D)`.
Obviously you need to replace the symbols above with the actual details.

>  hope this clarifies the motivation and justification.

It just discusses the solution you arrived at, it does not clearly describe the problem you were trying to solve.

Reply:
I was debugging this code with Juri yesterday and as the original author of the defrag method, I confirm that the offending line `tempDoc.copyOf(this, doc, doc);` is not only unnecessary here, but can cause database corruption. The `tempDoc` is only used as a temporary container for copying nodes during defrag and is thrown away afterwards! Original permissions and ownership on the resource are not affected by the `defragXMLResource` operation in any way, because the `tempDoc` is never written. What is written is the original DocumentImpl, which still has the proper permissions and ownership.

It seems I originally added the call to `tempDoc.copyOf` 19 years ago and it was preserved during a series of refactorings later though the implementation of `DocumentImpl` had changed to the point that it now may actually cause a corruption when called in this case.

Please be aware that this issue is a ticking time bomb that will lead to data wipe out!!! Together with the even more fatal #5296 (because it's easier to run into) this is absolutely worth a hotfix, so I would give it a thumbs up, not excluding that more improvements could be done at a later point.

Reply:
Hi @wolfgangmm thanks for your reply. It is interesting to hear about these symptoms, and it does sound like something needs to be fixed. However, the actual original issue and its cause remain unclear. I would like further information please so that we may all understand it...

> but can cause database corruption

Could you explain how that is possible please? As I understood it from @line-o's description there was a `PermissionDeniedException` which has not yet been explained. I am not sure how a `PermissionDeniedException`, which is a genuine way of forbidding a subject from doing something they are not allowed to do, could cause a database corruption here? 

> Original permissions and ownership on the resource are not affected by the defragXMLResource operation in any way, because the tempDoc is never written.

I did not expect that they would be.

> What is written is the original DocumentImpl, which still has the proper permissions and ownership.

Okay.

What I (and presumably others) don't yet understand, and what has not yet been explained, is how can the current approach that places the same permissions onto `tempDoc` as the original resource that is to be defragmented, cause a `PermissionDeniedException` (and then also if I now understand you, a subsequent database corruption)?

> Please be aware that this issue is a ticking time bomb that will lead to data wipe out!!!

I agree that it sounds like it needs to be fixed. At the moment I don't have enough information to see if the proposed fix correctly addresses the cause. I trust you and @line-o will provide the requested information about the underlying issue and cause...

> Together with the even more fatal https://github.com/eXist-db/exist/pull/5296 (because it's easier to run into)

Let's consider that separately. As at the moment https://github.com/eXist-db/exist/pull/5296 is blocked by work that is still needed to address issues introduced in https://github.com/eXist-db/exist/pull/5276 which itself was rushed and merged in violation of the agreed community process for reviewing and merging PRs.

> this is absolutely worth a hotfix,

From what you have said so far, I agree. We just need more information to be able to complete the review of this PR.

Reply:
@line-o added some information about how to reproduce the issue. It requires a certain combination of permissions. The corruption occurs because the PermissionDeniedException aborts the defrag process after some steps have already been applied. Therefore another way to prevent this might be to move the offending code further up to the start of the method. We did not follow this route as dropping the line is the cleaner way to solve the issue.

The PermissionDeniedException is eventually thrown in line 664 of `DocumentImpl#copyOf`: 

```
permissions.setOwner(prev.get_1().getOwner());
```

why this happens I cannot explain entirely and it seems very weird that the owners would be different. It must be related to the setgid. But in itself it has nothing to do with the actual defrag routine and points to another bug in permission handling.

Reply:
> it seems very weird that the owners would be different

Yes that was my feeling from the start, which is why I wanted more information from yourselves. Based on that knowledge it seems to me that the this PR is a workaround and not a fix for the root issue.

If @line-o could provide the information I previously requested [here](https://github.com/eXist-db/exist/pull/5311#discussion_r1612355788) please:

> Can you clearly describe the initial problem that you are trying to solve in the PR description please. For example, something like: There is an active user U with the security profile s(U), there is document D with the permissions p(D), when the action A is executed by the subject AS with security profile s(AS), it is rejected because s(AS) is incompatible with p(D).
Obviously you need to replace the symbols above with the actual details.

Then I should be able to understand, and then help you both resolve the real issue instead...

Reply:
@wolfgangmm I think you have accidentally deleted or overwritten my comments
<img width="541" alt="Screenshot 2024-05-24 at 16 39 27" src="https://github.com/eXist-db/exist/assets/1264057/e8a78d29-e4ae-4bbc-8ec6-4de2b7b46203">

Are you able to restore my comments please? If not, I can try and re-create them...